File: ios/App/Sources/Core/Domain/NavCaddy/Normalizer/GolfSlangDictionary.swift
import Foundation

/// Dictionary of golf slang terms and their normalized forms.
///
/// Spec R2: Input normalization - golf slang handling.
enum GolfSlangDictionary {

    // MARK: - Club Abbreviations

    /// Maps club abbreviations to full names.
    static let clubAbbreviations: [String: String] = [
        // Irons
        "1i": "1-iron",
        "2i": "2-iron",
        "3i": "3-iron",
        "4i": "4-iron",
        "5i": "5-iron",
        "6i": "6-iron",
        "7i": "7-iron",
        "8i": "8-iron",
        "9i": "9-iron",

        // Wedges
        "pw": "pitching wedge",
        "gw": "gap wedge",
        "aw": "approach wedge",
        "sw": "sand wedge",
        "lw": "lob wedge",

        // Woods
        "d": "driver",
        "dr": "driver",
        "1w": "driver",
        "2w": "2-wood",
        "3w": "3-wood",
        "4w": "4-wood",
        "5w": "5-wood",
        "7w": "7-wood",

        // Hybrids
        "3h": "3-hybrid",
        "4h": "4-hybrid",
        "5h": "5-hybrid",
        "6h": "6-hybrid",
        "7h": "7-hybrid"
    ]

    // MARK: - Common Golf Terms

    /// Maps slang terms to standard golf terminology.
    static let commonTerms: [String: String] = [
        "stick": "club",
        "sticks": "clubs",
        "the dance floor": "green",
        "dance floor": "green",
        "the short stuff": "fairway",
        "short stuff": "fairway",
        "the big stick": "driver",
        "big stick": "driver",
        "flatstick": "putter",
        "flat stick": "putter",
        "texas wedge": "putter",
        "worm burner": "low shot",
        "skull": "thin shot",
        "chunk": "fat shot",
        "duff": "fat shot",
        "shank": "toe shot",
        "banana ball": "slice",
        "duck hook": "severe hook",
        "snap hook": "severe hook",
        "pull": "pull shot",
        "push": "push shot",
        "fade": "controlled slice",
        "draw": "controlled hook",
        "dead straight": "straight",
        "pin high": "correct distance",
        "pin": "flagstick",
        "flag": "flagstick",
        "jar": "hole",
        "cup": "hole",
        "gimmie": "very short putt",
        "gimme": "very short putt",
        "birdie": "one under par",
        "eagle": "two under par",
        "albatross": "three under par",
        "bogey": "one over par",
        "double": "double bogey",
        "triple": "triple bogey",
        "snowman": "score of 8",
        "fried egg": "buried lie in sand",
        "beach": "sand bunker",
        "cabbage": "thick rough",
        "heavy rough": "thick rough",
        "up and down": "chip and one putt",
        "pitch and putt": "short course",
        "links": "seaside course",
        "track": "golf course",
        "layout": "golf course",
        "turn": "after 9 holes",
        "front nine": "holes 1-9",
        "back nine": "holes 10-18",
        "card": "scorecard",
        "tee it up": "start round",
        "first tee": "starting hole"
    ]

    // MARK: - Number Words

    /// Maps written numbers to digits.
    static let numberWords: [String: String] = [
        // Basic digits
        "zero": "0",
        "one": "1",
        "two": "2",
        "three": "3",
        "four": "4",
        "five": "5",
        "six": "6",
        "seven": "7",
        "eight": "8",
        "nine": "9",

        // Teens
        "ten": "10",
        "eleven": "11",
        "twelve": "12",
        "thirteen": "13",
        "fourteen": "14",
        "fifteen": "15",
        "sixteen": "16",
        "seventeen": "17",
        "eighteen": "18",
        "nineteen": "19",

        // Tens
        "twenty": "20",
        "thirty": "30",
        "forty": "40",
        "fifty": "50",
        "sixty": "60",
        "seventy": "70",
        "eighty": "80",
        "ninety": "90",

        // Hundreds
        "hundred": "100",
        "one hundred": "100",
        "two hundred": "200",
        "three hundred": "300"
    ]

    // MARK: - Composite Number Patterns

    /// Patterns for composite numbers like "one fifty" -> "150".
    static let compositePatterns: [(pattern: NSRegularExpression, transform: (String) -> String)] = [
        // Pattern: "one fifty" -> "150"
        (
            pattern: try! NSRegularExpression(pattern: #"\b(one|two|three) (ten|twenty|thirty|forty|fifty|sixty|seventy|eighty|ninety)\b"#, options: .caseInsensitive),
            transform: { match in
                let components = match.lowercased().split(separator: " ").map(String.init)
                guard components.count == 2 else { return match }

                let hundreds: [String: String] = ["one": "1", "two": "2", "three": "3"]
                let tens: [String: String] = ["ten": "10", "twenty": "20", "thirty": "30", "forty": "40", "fifty": "50", "sixty": "60", "seventy": "70", "eighty": "80", "ninety": "90"]

                guard let h = hundreds[components[0]], let t = tens[components[1]] else { return match }
                return h + t
            }
        ),

        // Pattern: "twenty three" -> "23"
        (
            pattern: try! NSRegularExpression(pattern: #"\b(twenty|thirty|forty|fifty|sixty|seventy|eighty|ninety)[\s-](one|two|three|four|five|six|seven|eight|nine)\b"#, options: .caseInsensitive),
            transform: { match in
                let normalized = match.lowercased().replacingOccurrences(of: "-", with: " ")
                let components = normalized.split(separator: " ").map(String.init)
                guard components.count == 2 else { return match }

                let tens: [String: String] = ["twenty": "2", "thirty": "3", "forty": "4", "fifty": "5", "sixty": "6", "seventy": "7", "eighty": "8", "ninety": "9"]
                let ones: [String: String] = ["one": "1", "two": "2", "three": "3", "four": "4", "five": "5", "six": "6", "seven": "7", "eight": "8", "nine": "9"]

                guard let t = tens[components[0]], let o = ones[components[1]] else { return match }
                return t + o
            }
        ),

        // Pattern: "seven iron" -> "7-iron"
        (
            pattern: try! NSRegularExpression(pattern: #"\b(one|two|three|four|five|six|seven|eight|nine)\s+iron\b"#, options: .caseInsensitive),
            transform: { match in
                let components = match.lowercased().split(separator: " ").map(String.init)
                guard components.count == 2 else { return match }

                let numbers: [String: String] = ["one": "1", "two": "2", "three": "3", "four": "4", "five": "5", "six": "6", "seven": "7", "eight": "8", "nine": "9"]

                guard let num = numbers[components[0]] else { return match }
                return num + "-iron"
            }
        ),

        // Pattern: "three wood" -> "3-wood"
        (
            pattern: try! NSRegularExpression(pattern: #"\b(one|two|three|four|five|six|seven)\s+wood\b"#, options: .caseInsensitive),
            transform: { match in
                let components = match.lowercased().split(separator: " ").map(String.init)
                guard components.count == 2 else { return match }

                let numbers: [String: String] = ["one": "1", "two": "2", "three": "3", "four": "4", "five": "5", "six": "6", "seven": "7"]

                guard let num = numbers[components[0]] else { return match }
                return num + "-wood"
            }
        ),

        // Pattern: "four hybrid" -> "4-hybrid"
        (
            pattern: try! NSRegularExpression(pattern: #"\b(one|two|three|four|five|six|seven)\s+hybrid\b"#, options: .caseInsensitive),
            transform: { match in
                let components = match.lowercased().split(separator: " ").map(String.init)
                guard components.count == 2 else { return match }

                let numbers: [String: String] = ["one": "1", "two": "2", "three": "3", "four": "4", "five": "5", "six": "6", "seven": "7"]

                guard let num = numbers[components[0]] else { return match }
                return num + "-hybrid"
            }
        )
    ]

    // MARK: - Profanity Filter

    /// Common profanity patterns (replace with asterisks or remove).
    /// Note: This is a minimal set for demonstration. A production app should use
    /// a comprehensive profanity filter service or library.
    static let profanityPatterns: [String] = [
        "damn",
        "hell",
        "crap",
        "shit",
        "fuck",
        "ass",
        "bitch",
        "bastard",
        "piss"
    ]

    /// Create regex pattern for profanity filtering.
    static func profanityRegex() -> NSRegularExpression? {
        let pattern = profanityPatterns
            .map { NSRegularExpression.escapedPattern(for: $0) }
            .joined(separator: "|")

        let fullPattern = #"\b(\#(pattern))\w*\b"#
        return try? NSRegularExpression(pattern: fullPattern, options: .caseInsensitive)
    }
}

File: android/app/src/main/java/com/example/app/domain/navcaddy/normalizer/GolfSlangDictionary.kt
package caddypro.domain.navcaddy.normalizer

/**
 * Dictionary of golf slang and common terminology.
 *
 * Spec reference: navcaddy-engine.md R2, navcaddy-engine-plan.md Task 7
 *
 * Mappings include:
 * - Club abbreviations (7i, PW, SW, etc.)
 * - Woods and driver abbreviations (3w, 5w, D)
 * - Hybrid abbreviations (3h, 4h)
 * - Common golf terms ("stick" -> "club", "dance floor" -> "green")
 * - Number words ("one" -> "1", "fifty" -> "50")
 */
object GolfSlangDictionary {

    /**
     * Club abbreviations to full names.
     * Case-insensitive matching is applied by the normalizer.
     */
    val clubAbbreviations = mapOf(
        // Irons with 'i' suffix
        "3i" to "3-iron",
        "4i" to "4-iron",
        "5i" to "5-iron",
        "6i" to "6-iron",
        "7i" to "7-iron",
        "8i" to "8-iron",
        "9i" to "9-iron",

        // Wedges
        "pw" to "pitching wedge",
        "gw" to "gap wedge",
        "aw" to "approach wedge",
        "sw" to "sand wedge",
        "lw" to "lob wedge",

        // Woods with 'w' suffix
        "3w" to "3-wood",
        "5w" to "5-wood",
        "7w" to "7-wood",

        // Driver
        "d" to "driver",

        // Hybrids with 'h' suffix
        "2h" to "2-hybrid",
        "3h" to "3-hybrid",
        "4h" to "4-hybrid",
        "5h" to "5-hybrid"
    )

    /**
     * Common golf terms and slang.
     */
    val commonTerms = mapOf(
        "stick" to "club",
        "sticks" to "clubs",
        "dance floor" to "green",
        "the dance floor" to "green",
        "tin cup" to "hole",
        "bunker" to "sand trap",
        "putting surface" to "green",
        "fairway metal" to "fairway wood",
        "big stick" to "driver",
        "big dog" to "driver",
        "flat stick" to "putter"
    )

    /**
     * Number words to digits for yardages and scores.
     * These handle basic number conversions.
     */
    val numberWords = mapOf(
        // Single digits
        "one" to "1",
        "two" to "2",
        "three" to "3",
        "four" to "4",
        "five" to "5",
        "six" to "6",
        "seven" to "7",
        "eight" to "8",
        "nine" to "9",

        // Teens
        "ten" to "10",
        "eleven" to "11",
        "twelve" to "12",
        "thirteen" to "13",
        "fourteen" to "14",
        "fifteen" to "15",
        "sixteen" to "16",
        "seventeen" to "17",
        "eighteen" to "18",
        "nineteen" to "19",

        // Tens
        "twenty" to "20",
        "thirty" to "30",
        "forty" to "40",
        "fifty" to "50",
        "sixty" to "60",
        "seventy" to "70",
        "eighty" to "80",
        "ninety" to "90",

        // Hundreds
        "hundred" to "100",
        "hundred and" to "100"
    )

    /**
     * Compound number patterns (e.g., "one fifty" -> "150").
     * These are ordered by specificity (most specific first).
     */
    val compoundNumbers = listOf(
        // Hundred patterns (e.g., "one hundred fifty" -> "150")
        "one hundred fifty" to "150",
        "one hundred sixty" to "160",
        "one hundred seventy" to "170",
        "one hundred eighty" to "180",
        "one hundred ninety" to "190",
        "two hundred" to "200",

        // Shortened hundred patterns (e.g., "one fifty" -> "150")
        "one fifty" to "150",
        "one sixty" to "160",
        "one seventy" to "170",
        "one eighty" to "180",
        "one ninety" to "190",
        "two hundred" to "200",

        // Common yardage patterns
        "one twenty" to "120",
        "one thirty" to "130",
        "one forty" to "140",
        "one ten" to "110",

        // Twenty patterns (e.g., "twenty five" -> "25")
        "twenty one" to "21",
        "twenty two" to "22",
        "twenty three" to "23",
        "twenty four" to "24",
        "twenty five" to "25",
        "twenty six" to "26",
        "twenty seven" to "27",
        "twenty eight" to "28",
        "twenty nine" to "29",

        // Thirty patterns
        "thirty one" to "31",
        "thirty two" to "32",
        "thirty three" to "33",
        "thirty four" to "34",
        "thirty five" to "35",
        "thirty six" to "36",
        "thirty seven" to "37",
        "thirty eight" to "38",
        "thirty nine" to "39",

        // Iron patterns (e.g., "seven iron" -> "7-iron")
        "three iron" to "3-iron",
        "four iron" to "4-iron",
        "five iron" to "5-iron",
        "six iron" to "6-iron",
        "seven iron" to "7-iron",
        "eight iron" to "8-iron",
        "nine iron" to "9-iron",

        // Wood patterns
        "three wood" to "3-wood",
        "five wood" to "5-wood",
        "seven wood" to "7-wood"
    )

    /**
     * Get all mappings as a single map for easy lookup.
     */
    fun getAllMappings(): Map<String, String> {
        return clubAbbreviations + commonTerms + numberWords
    }
}
